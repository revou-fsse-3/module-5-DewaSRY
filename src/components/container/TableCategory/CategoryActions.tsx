import { FC, HTMLAttributes, PropsWithChildren } from "react";
import Button from "@common/Button";
import { useNavigate } from "react-router-dom";
type CategoryActionsProps = HTMLAttributes<HTMLDivElement> & {
  id: string;
};
type CategoryActionsComponents = FC<CategoryActionsProps> & PropsWithChildren;
const CategoryActions: CategoryActionsComponents = ({ id }) => {
  const navigation = useNavigate();
  const handleActions = () => {
    navigation("update?id=" + id);
  };
  return <Button onClick={handleActions}>Actions</Button>;
};

export default CategoryActions;
