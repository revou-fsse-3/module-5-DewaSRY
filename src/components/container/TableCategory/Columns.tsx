import { createColumnHelper } from "@tanstack/react-table";
import { CollectionsProps } from "@features/utils/collections.type";
import Link from "next/link";
const columnHelper = createColumnHelper<CollectionsProps>();
const columns = [
  columnHelper.accessor("name", {
    cell: (props) => <span>{props.getValue()}</span>,
    header: () => <span>Last Name</span>,
    footer: (props) => props.column.id,
  }),
  columnHelper.accessor("is_active", {
    cell: (props) => <span>{props.getValue() ?? "active"}</span>,
    header: () => <span>Active Status</span>,
    footer: (props) => props.column.id,
  }),
  columnHelper.accessor("created_at", {
    cell: (props) => {
      const date = new Date(props.getValue());
      const formatData = formattedDate(date);
      return <span>{formatData}</span>;
    },
    header: () => <span>Create</span>,
    footer: (props) => props.column.id,
  }),
  columnHelper.accessor("updated_at", {
    cell: (props) => {
      const date = new Date(props.getValue());
      const formatData = formattedDate(date);
      return <span>{formatData}</span>;
    },
    header: () => <span>Update</span>,
    footer: (props) => props.column.id,
  }),
  columnHelper.accessor("id", {
    cell: (props) => (
      <Link href={`/collection/update?id=${props.getValue()}`}>Actions</Link>
    ),
    header: () => <span>Actions</span>,
    footer: (props) => props.column.id,
  }),
];

export default columns;
const formattedDate = (date: Date) => {
  return date.toLocaleDateString("en-US", {
    day: "numeric",
    month: "short",
    year: "numeric",
  });
};
