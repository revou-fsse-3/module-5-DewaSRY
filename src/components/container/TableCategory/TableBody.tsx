import { FC, HTMLAttributes, PropsWithChildren } from "react";
import { flexRender, Table } from "@tanstack/react-table";
import { CollectionsProps } from "@features/utils/collections.type";
interface TableBodyProps extends HTMLAttributes<HTMLTableSectionElement> {
  table: Table<CollectionsProps>;
}
type TableBodyComponents = FC<TableBodyProps> & PropsWithChildren;
const TableBody: TableBodyComponents = ({ table, ...resProps }) => {
  const TABLE_ROWS = table.getRowModel().rows;
  return (
    <tbody
      {...resProps}
      className={`${resProps.className ? resProps.className : ""}`}
    >
      {TABLE_ROWS.length ? (
        TABLE_ROWS.map((row, i) => (
          <tr
            key={row.id}
            className={`
                ${i % 2 === 0 ? "bg-gray-100" : "bg-gray-200"}
                `}
          >
            <td>{i + 1} </td>
            {row.getVisibleCells().map((cell, id) => (
              <td key={cell.id + id} className="px-3.5 py-2">
                {flexRender(cell.column.columnDef.cell, cell.getContext())}
              </td>
            ))}
          </tr>
        ))
      ) : (
        <tr className="text-center h-32">
          <td colSpan={12}>No Record Found!</td>
        </tr>
      )}
    </tbody>
  );
};

export default TableBody;
