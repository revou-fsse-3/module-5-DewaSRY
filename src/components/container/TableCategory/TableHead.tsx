import { FC, HTMLAttributes, PropsWithChildren, Fragment } from "react";
import { flexRender, Table } from "@tanstack/react-table";
import { CollectionsProps } from "@features/utils/collections.type";

interface TableHeadProps extends HTMLAttributes<HTMLTableSectionElement> {
  table: Table<CollectionsProps>;
}
type TableHeadComponents = FC<TableHeadProps> & PropsWithChildren;
const TableHead: TableHeadComponents = ({ table, ...resProps }) => {
  return (
    <thead
      {...resProps}
      className={`${resProps.className ? resProps.className : ""}`}
    >
      {table.getHeaderGroups().map((headerGroup) => (
        <tr key={headerGroup.id}>
          <th>NO</th>
          {headerGroup.headers.map((header, id) => {
            return (
              <th key={header.id + id} colSpan={header.colSpan}>
                {header.isPlaceholder ? null : (
                  <div
                    {...{
                      className: header.column.getCanSort()
                        ? "cursor-pointer select-none"
                        : "",
                      onClick: header.column.getToggleSortingHandler(),
                    }}
                  >
                    <Fragment>
                      {flexRender(
                        header.column.columnDef.header,
                        header.getContext()
                      )}
                    </Fragment>
                    <Fragment>
                      {{
                        asc: " ðŸ”¼",
                        desc: " ðŸ”½",
                      }[header.column.getIsSorted() as string] ?? null}
                    </Fragment>
                  </div>
                )}
              </th>
            );
          })}
        </tr>
      ))}
    </thead>
  );
};

export default TableHead;
