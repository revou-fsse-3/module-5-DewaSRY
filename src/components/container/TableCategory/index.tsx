import { useState } from "react";
import { FC, TableHTMLAttributes, PropsWithChildren } from "react";
import COLUMNS from "./Columns";
import TableHead from "./TableHead";
import TableBody from "./TableBody";
import { CollectionsProps } from "@features/utils/collections.type";
import {
  getCoreRowModel,
  getSortedRowModel,
  SortingState,
  useReactTable,
} from "@tanstack/react-table";

interface TableProps extends TableHTMLAttributes<HTMLTableElement> {
  data: CollectionsProps[];
}
type TableComponents = FC<TableProps> & PropsWithChildren;
const Table: TableComponents = ({ data, ...resProps }) => {
  const [sorting, setSorting] = useState<SortingState>([]);
  const table = useReactTable<CollectionsProps>({
    data,
    columns: COLUMNS,
    state: {
      sorting,
    },
    onSortingChange: setSorting,
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    debugTable: true,
  });
  return (
    <table
      {...resProps}
      className={`${resProps.className ? resProps.className : ""}`}
    >
      <TableHead table={table} />
      <TableBody table={table} />
    </table>
  );
};

export default Table;
