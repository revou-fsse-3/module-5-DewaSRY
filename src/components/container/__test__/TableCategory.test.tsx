import { describe, expect, vi, it, beforeAll, beforeEach } from "vitest";
import { render, fireEvent, screen, act, waitFor } from "@libs/reactTesting";
import "@testing-library/jest-dom";
import TableCategory from "@container/TableCategory";
import collectionRespond from "@libs/mock/collections/AllCollections";

describe("testing Table Category container", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    render(<TableCategory data={collectionRespond.data} role="main" />);
  });

  it("table Category match with snapshot ", () => {
    const main = screen.getAllByRole("main");
    expect(main).toMatchSnapshot();
  });
});
