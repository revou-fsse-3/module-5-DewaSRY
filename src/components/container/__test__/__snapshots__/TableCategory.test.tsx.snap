// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`testing Table Category container > table Category match with snapshot  1`] = `
[
  <table
    class=""
    role="main"
  >
    <thead
      class=""
    >
      <tr>
        <th>
          NO
        </th>
        <th
          colspan="1"
        >
          <div
            class="cursor-pointer select-none"
          >
            <span>
              Last Name
            </span>
          </div>
        </th>
        <th
          colspan="1"
        >
          <div
            class="cursor-pointer select-none"
          >
            <span>
              Active Status
            </span>
          </div>
        </th>
        <th
          colspan="1"
        >
          <div
            class="cursor-pointer select-none"
          >
            <span>
              Create
            </span>
          </div>
        </th>
        <th
          colspan="1"
        >
          <div
            class="cursor-pointer select-none"
          >
            <span>
              Update
            </span>
          </div>
        </th>
        <th
          colspan="1"
        >
          <div
            class="cursor-pointer select-none"
          >
            <span>
              Actions
            </span>
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class="
                bg-gray-100
                "
      >
        <td>
          1
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            mock halloo apa kabar  loo gory
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=32022cb9-eb8f-4f91-90d5-ddb49b4cb3f0"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-200
                "
      >
        <td>
          2
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            mock halloo apa kabar  loo gory one two
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=6a78f162-af26-4c3d-b9bb-acd5b9534b84"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-100
                "
      >
        <td>
          3
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            mock halloo apa kabar  loo gory one
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=8ffd60fb-538a-4115-83e5-6dc1b9636681"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-200
                "
      >
        <td>
          4
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            asdasfasdasdasdasdafasfasfasfd
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=e008d82c-4cea-4e6c-a110-31c12838366b"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-100
                "
      >
        <td>
          5
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            before Update
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=ec8874da-d514-4a48-a1b4-1def81d6537f"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-200
                "
      >
        <td>
          6
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            before Update
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=ec8874da-d514-4a48-a1b4-1def81d65hhf"
          >
            Actions
          </a>
        </td>
      </tr>
      <tr
        class="
                bg-gray-100
                "
      >
        <td>
          7
           
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            testing create
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span />
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <span>
            Jan 25, 2024
          </span>
        </td>
        <td
          class="px-3.5 py-2"
        >
          <a
            href="/collection/update?id=e01c57f7-a0b3-4704-9ce4-57aa14a9d401"
          >
            Actions
          </a>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
